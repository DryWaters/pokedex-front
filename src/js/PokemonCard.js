import React from 'react';
import PropTypes from 'prop-types';

import PokemonType from './PokemonType';

import '../styles/PokemonCard.css';

const baseURL = 'https://pokedex-backend-server.herokuapp.com';

const propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  types: PropTypes.arrayOf(PropTypes.string).isRequired,
  image_path: PropTypes.shape({
  small: PropTypes.string.isRequired,
  large: PropTypes.string.isRequired,
  }).isRequired,
};

class PokemonCard extends React.Component {

  /**
   * Capitalizes a name.
   * The first letter should be capitalized, as well as any letter following a space or dash
   * @param  {String} name - the name to capitalize
   * @return {String} - the capitalized name
   */
  capitalizeName (name) {
    // Capitalize first letter
    let capitalizedLetter = name.charAt(0).toUpperCase();
    let capitalizedName = capitalizedLetter + name.slice(1);

    // Capitalize other letters that should be capitalized.
    let i = capitalizedName.search(/[- ]/);
    while (i > -1) {
      capitalizedLetter = capitalizedName.charAt(i + 1).toUpperCase();
      capitalizedName = capitalizedName.slice(0, i + 1) + capitalizedLetter + capitalizedName.slice(i + 2);

      let next = capitalizedName.substring(i + 1).search(/[- ]/);
      i = next > -1 ? i + 1 + next : -1;
    }

    return capitalizedName;
  }

  render () {
    return (
      <div className='card' >
        <img src={baseURL + this.props.image_path.large} alt={this.props.name + ' sprite'} />
        <div className="textarea" align="center" >
          <h5 className="pokemon-id" align="left">{'#' + this.props.id}</h5>
          <h3><b>{this.capitalizeName(this.props.name)}</b></h3>
          <PokemonType id={this.props.id} types= {this.props.types} />
        </div>
      </div>
    );
  }
}

PokemonCard.propTypes = propTypes;

export default PokemonCard;
