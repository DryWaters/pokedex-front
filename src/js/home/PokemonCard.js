import React from 'react';
import PropTypes from 'prop-types';

import PokemonType from '../core/PokemonType';
import { getPokemonSprite } from '../core/PokemonAPI';

import '../../styles/home/PokemonCard.css';

const propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  types: PropTypes.arrayOf(PropTypes.string).isRequired,
  image_path: PropTypes.string.isRequired,
  handlePokemonCardClick: PropTypes.func.isRequired
};

class PokemonCard extends React.Component {

  render () {
    return (
      <div className='card' onClick={() => {this.props.handlePokemonCardClick(this.props.id)}} >
        <img className="card-sprite" src={getPokemonSprite(this.props.image_path)} alt={`${this.props.name} sprite`} />
        <div className="textarea" align="center" >
          <h5 className="pokemon-id" align="left">{`#${this.props.id}`}</h5>
          <h3><b>{this.props.name}</b></h3>
          <PokemonType isModal={false} isWeakness={false} id={this.props.id} types={this.props.types} />
        </div>
      </div>
    );
  }
}

PokemonCard.propTypes = propTypes;

export default PokemonCard;
