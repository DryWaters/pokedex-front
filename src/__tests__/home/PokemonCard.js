import React from 'react';
import { shallow } from 'enzyme';

import PokemonCard from '../../js/home/PokemonCard';
import ExampleJSON from '../../../local_data/example_single.json';

describe('PokemonCard', () => {
  it('should render Card component without crashing', () => {
    shallow(<PokemonCard handlePokemonCardClick={() => {}} {...ExampleJSON} />);
  });

  it('should match Card component snapshot', () => {
    const wrapper = shallow(<PokemonCard handlePokemonCardClick={() => {}} {...ExampleJSON} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('capitalizeName', () => {
    it('should do nothing to an empty string', () => {
      const emptyStr = PokemonCard.prototype.capitalizeName('');
      expect(emptyStr).toEqual('');
    });

    it('should capitalize the first letter of a string', () => {
      const capitalizedName = PokemonCard.prototype.capitalizeName('bulbasaur');
      expect(capitalizedName).toEqual('Bulbasaur');
    })

    it('should capitalize the letter after a space', () => {
      const capitalizedName = PokemonCard.prototype.capitalizeName('ash ketchum');
      expect(capitalizedName).toEqual('Ash Ketchum');
    })

    it('should capitalize the letter after a dash', () => {
      const capitalizedName = PokemonCard.prototype.capitalizeName('porygon-z');
      expect(capitalizedName).toEqual('Porygon-Z');
    });

    it('should handle multiple spaces', () => {
      const capitalizedName = PokemonCard.prototype.capitalizeName('bug catcher greg');
      expect(capitalizedName).toEqual('Bug Catcher Greg');
    });

    it('should handle multiple dashes', () => {
      const capitalizedName = PokemonCard.prototype.capitalizeName('porygon-z-again');
      expect(capitalizedName).toEqual('Porygon-Z-Again');
    });

    it('should handle a mix of spaces and dashes', () => {
      const capitalizedName = PokemonCard.prototype.capitalizeName('porygon-z again');
      expect(capitalizedName).toEqual('Porygon-Z Again');
    });
  });
});
