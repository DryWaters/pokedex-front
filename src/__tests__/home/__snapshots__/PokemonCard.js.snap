// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard should match Card component snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PokemonCard
    handlePokemonCardClick={[Function]}
    id={130}
    image_path="./130.jpg"
    name="gyarados"
    types={
        Array [
            "water",
            "flying",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="gyarados sprite"
          className="card-sprite"
          src="https://pokedex-backend-server.herokuapp.com./130.jpg"
/>,
        <div
          align="center"
          className="textarea"
>
          <h5
                    align="left"
                    className="pokemon-id"
          >
                    #130
          </h5>
          <h3>
                    <b>
                              Gyarados
                    </b>
          </h3>
          <PokemonType
                    id={130}
                    isModal={false}
                    isWeakness={false}
                    types={
                              Array [
                                        "water",
                                        "flying",
                                      ]
                    }
          />
</div>,
      ],
      "className": "card",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "gyarados sprite",
          "className": "card-sprite",
          "src": "https://pokedex-backend-server.herokuapp.com./130.jpg",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "align": "center",
          "children": Array [
            <h5
              align="left"
              className="pokemon-id"
>
              #130
</h5>,
            <h3>
              <b>
                            Gyarados
              </b>
</h3>,
            <PokemonType
              id={130}
              isModal={false}
              isWeakness={false}
              types={
                            Array [
                                          "water",
                                          "flying",
                                        ]
              }
/>,
          ],
          "className": "textarea",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "align": "left",
              "children": "#130",
              "className": "pokemon-id",
            },
            "ref": null,
            "rendered": "#130",
            "type": "h5",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <b>
                Gyarados
</b>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Gyarados",
              },
              "ref": null,
              "rendered": "Gyarados",
              "type": "b",
            },
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "id": 130,
              "isModal": false,
              "isWeakness": false,
              "types": Array [
                "water",
                "flying",
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="gyarados sprite"
            className="card-sprite"
            src="https://pokedex-backend-server.herokuapp.com./130.jpg"
/>,
          <div
            align="center"
            className="textarea"
>
            <h5
                        align="left"
                        className="pokemon-id"
            >
                        #130
            </h5>
            <h3>
                        <b>
                                    Gyarados
                        </b>
            </h3>
            <PokemonType
                        id={130}
                        isModal={false}
                        isWeakness={false}
                        types={
                                    Array [
                                                "water",
                                                "flying",
                                              ]
                        }
            />
</div>,
        ],
        "className": "card",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "gyarados sprite",
            "className": "card-sprite",
            "src": "https://pokedex-backend-server.herokuapp.com./130.jpg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "align": "center",
            "children": Array [
              <h5
                align="left"
                className="pokemon-id"
>
                #130
</h5>,
              <h3>
                <b>
                                Gyarados
                </b>
</h3>,
              <PokemonType
                id={130}
                isModal={false}
                isWeakness={false}
                types={
                                Array [
                                                "water",
                                                "flying",
                                              ]
                }
/>,
            ],
            "className": "textarea",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "align": "left",
                "children": "#130",
                "className": "pokemon-id",
              },
              "ref": null,
              "rendered": "#130",
              "type": "h5",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <b>
                  Gyarados
</b>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Gyarados",
                },
                "ref": null,
                "rendered": "Gyarados",
                "type": "b",
              },
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "id": 130,
                "isModal": false,
                "isWeakness": false,
                "types": Array [
                  "water",
                  "flying",
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
